examples/carousel/js/carouselTab.js

```diff
@@ -0,0 +1,108 @@
+/*
+*   File:   carouselTab.js
+*
+*   Desc:   Carousel Tab widget that implements ARIA Authoring Practices
+*/
+
+/*
+*   @constructor CarouselTab
+*
+*
+*/
+var CarouselTab = function (domNode, carouselObj, tabpanelObj) {
+  this.domNode = domNode;
+  this.carousel = carouselObj;
+  this.tabpanel = tabpanelObj;
+
+  this.keyCode = Object.freeze({
+    'RETURN': 13,
+    'SPACE': 32,
+    'END': 35,
+    'HOME': 36,
+    'LEFT': 37,
+    'UP': 38,
+    'RIGHT': 39,
+    'DOWN': 40
+  });
+};
+
+CarouselTab.prototype.init = function () {
+  this.domNode.addEventListener('keydown', this.handleKeydown.bind(this));
+  this.domNode.addEventListener('click', this.handleClick.bind(this));
+  this.domNode.addEventListener('focus', this.handleFocus.bind(this));
+  this.domNode.addEventListener('blur', this.handleBlur.bind(this));
+};
+
+CarouselTab.prototype.hideTabpanel = function () {
+  this.domNode.setAttribute('aria-selected', 'false');
+  this.tabpanel.hide();
+  this.domNode.setAttribute('tabindex', '-1');
+};
+
+CarouselTab.prototype.showTabpanel = function (moveFocus) {
+  if (typeof moveFocus !== 'boolean') {
+    moveFocus = false;
+  }
+
+  this.domNode.setAttribute('aria-selected', 'true');
+  this.domNode.removeAttribute('tabindex');
+  this.tabpanel.show();
+  if (moveFocus) {
+    this.domNode.focus();
+  }
+};
+
+
+/* EVENT HANDLERS */
+
+
+CarouselTab.prototype.handleKeydown = function (event) {
+  var flag = false;
+
+  switch (event.keyCode) {
+
+    case this.keyCode.RIGHT:
+      this.carousel.setSelectedToNextTab(this, true);
+      flag = true;
+      break;
+
+    case this.keyCode.LEFT:
+      this.carousel.setSelectedToPreviousTab(this, true);
+      flag = true;
+      break;
+
+    case this.keyCode.HOME:
+      this.carousel.setSelectedToFirstTab(true);
+      flag = true;
+      break;
+
+    case this.keyCode.END:
+      this.carousel.setSelectedToLastTab(true);
+      flag = true;
+      break;
+
+    default:
+      break;
+  }
+
+  if (flag) {
+    event.stopPropagation();
+    event.preventDefault();
+  }
+};
+
+CarouselTab.prototype.handleClick = function () {
+  this.carousel.setSelectedTab(this, true);
+};
+
+CarouselTab.prototype.handleFocus = function () {
+  this.carousel.hasFocus = true;
```

Comment:
This seems like the kind of thing we could consensus on and put into coding best practices. Issue #1180 is called "code readability", but could it be expanded to best practices/norms for the javascript in the examples? I'm fine with either of the latter two. @ZoeBijl and @jongund, do you have a preference between this options?

Reply:
This is a perfect example of what should go into the Code Readability docs. In my opinion the second to last option is best. I often find myself wanting to split a component into more classes. But always keep it in one because I find that easier to read. Open to other opinions tho :).

Reply:
I'm for keeping them in one class, at least unless there's a sub-component that would truly be reusable across multiple componenst, e.g. a month spinbutton within a DatePicker dropdown. But I'd also be fine with that not being split out for the sake of cohesive examples, and because we're not building a production-style pattern library with shared code across components.

Reply:
Perhaps we could combine it into something like: “All of a widgets’ logic should be contained in a single class. If there’s a good reason to split it into multiple classes communication between them should be done via events.” But like with better words.

Your comment does speak to why I prefer the single class strategy; we don’t need this to be production code—in fact, it’s better if it isn’t.

Reply:
@jongund do you feel like you have enough guidance to do these changes? Thanks for your work so far and it's great we have your code to work out these norms with!

Reply:
@spectranaut I have been traveling the last week, so I am trying to catch up with all the comments on this example, but they seem to make sense so far.